lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1665,22 +1665,61 @@ static void diagRecursivePropertyAccess(TypeChecker &TC, const Expr *E,
 /// Within a closure, we require that the source code contain "self." explicitly
 /// because 'self' is captured, not the property value.  This is a common source
 /// of confusion, so we force an explicit self.
-static void diagnoseImplicitSelfUseInClosure(TypeChecker &TC, const Expr *E,
-                                             const DeclContext *DC) {
+///
+/// Also perform primitive checks for reference cycles and warn the user.
+static void diagnoseVarUseInClosure(TypeChecker &TC, const Expr *E,
+                                    const DeclContext *DC,
+                                    bool suppressCycleDiags) {
   class DiagnoseWalker : public ASTWalker {
     TypeChecker &TC;
     unsigned InClosure;
+    bool SuppressCycleDiags;
+
+    enum IsIndirectlyReferenced_t : bool {
+      IsDirectlyReferenced = false, IsIndirectlyReferenced = true
+    };
+    using ReferenceCycleRoot = llvm::PointerIntPair<VarDecl *, 1, IsIndirectlyReferenced_t>;
   public:
-    explicit DiagnoseWalker(TypeChecker &TC, bool isAlreadyInClosure)
-        : TC(TC), InClosure(isAlreadyInClosure) {}
+    explicit DiagnoseWalker(TypeChecker &TC,
+                            bool isAlreadyInClosure,
+                            bool suppressCycleDiags)
+      : TC(TC), InClosure(isAlreadyInClosure),
+        SuppressCycleDiags(suppressCycleDiags) {}
+
+    static bool isStronglyOwnedVariable(VarDecl *var) {
+      if (!var->getType()->hasReferenceSemantics())
+        return false;
 
-    /// Return true if this is an implicit reference to self.
-    static bool isImplicitSelfUse(Expr *E) {
+      if (auto *OA = var->getAttrs().getAttribute<ReferenceOwnershipAttr>())
+        if (OA->isValid() && OA->get() != ReferenceOwnership::Strong)
```

Comment:
`getAttribute` ignores invalid attributes by default, so you can skip this check.